fvm/state/state.go

```diff
@@ -79,132 +84,210 @@ func WithMaxInteractionSizeAllowed(limit uint64) func(st *State) *State {
+	// reduce the bytes to be written
+	if old, ok := s.delta[p.PayloadKey]; ok {
+		s.ToBeWrittenCounter--
+		s.TotalBytesToBeWritten -= old.size()
+	}
 
-	return value, s.updateInteraction(owner, controller, key, value, []byte{})
+	s.delta[p.PayloadKey] = *p
+	s.ToBeWrittenCounter++
+	s.TotalBytesToBeWritten += p.size()
 }
 
+// Set updates state delta with a register update
 func (s *State) Set(owner, controller, key string, value flow.RegisterValue) error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
 	if err := s.checkSize(owner, controller, key, value); err != nil {
 		return err
 	}
 
-	s.draft[fullKey(owner, controller, key)] = payload{owner, controller, key, value}
+	pKey := PayloadKey{owner, controller, key}
+	p := Payload{pKey, value}
+	s.LogTouch(&p)
+
+	s.updateDelta(&p)
+
 	address, isAddress := addressFromOwner(owner)
 	if isAddress {
 		s.updatedAddresses[address] = struct{}{}
 	}
 	return nil
 }
 
-func (s *State) Touch(owner, controller, key string) error {
-	_, err := s.Get(owner, controller, key)
-	return err
-}
-
 func (s *State) Delete(owner, controller, key string) error {
 	err := s.Set(owner, controller, key, nil)
 	return err
 }
 
-func (s *State) Commit() error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+// We don't need this later, it should be invisible to the cadence
+func (s *State) Touch(owner, controller, key string) error {
+	s.LogTouch(&Payload{PayloadKey{owner, controller, key}, nil})
+	return nil
+}
+
+// NewChild generates a new child state
+func (s *State) NewChild() *State {
+	return NewState(s.ledger,
+		WithParent(s),
+		WithMaxKeySizeAllowed(s.maxKeySizeAllowed),
+		WithMaxValueSizeAllowed(s.maxValueSizeAllowed),
+		WithMaxInteractionSizeAllowed(s.maxInteractionAllowed),
+	)
+}
+
+func (s *State) MergeTouchLogs(child *State) error {
+	// append touches
+	s.LogTouches(child.touchLog)
+	// TODO maybe merge read cache for performance on failed cases
+	return nil
+}
 
-	// TODO right now we sort keys to minimize
-	// the impact on the spock, but later we might
-	// expose interactionFingerprint as a separate
-	// parameters, this preserve number of reads and
-	// order of reads as extra layer of entropy
+// MergeAnyState applies the changes from any given state
+func (s *State) MergeAnyState(other *State) error {
+	// append touches
+	s.LogTouches(other.touchLog)
 
-	keys := make([]string, 0)
-	for k := range s.draft {
-		keys = append(keys, k)
+	// apply delta
+	for _, v := range other.delta {
+		s.updateDelta(&v)
 	}
 
-	sort.Strings(keys)
+	// apply address updates
+	for k, v := range other.updatedAddresses {
+		s.updatedAddresses[k] = v
+	}
 
-	for _, k := range keys {
-		p := s.draft[k]
-		oldValue, err := s.ledger.Get(p.owner, p.controller, p.key)
-		if err != nil {
-			return err
-		}
-		// update interaction
-		err = s.updateInteraction(p.owner, p.controller, p.key, oldValue, p.value)
+	// update ledger interactions
+	s.ReadCounter += other.ReadCounter
+	s.WriteCounter += other.WriteCounter
```

Comment:
Ok... I think I understand how the write counter works now... and is this line necessary?

Reply:
this one is necessary but toBeWritten won't be updated since the delta are not merged

Reply:
I removed the point of confusion, WriteCounter will be what you need.